File: ts/src/tokens.ts
/**
 * Token measurement — parity port of contexel's Python `tokens` module.
 *
 * Counting is encoding-relative: a record's cost is the token count of its
 * serialized text. The canonical serialization replicates Python's
 * `json.dumps(value, ensure_ascii=False, default=str, sort_keys=True)` —
 * including the ", " / ": " separators — so both languages price identical
 * records identically. Token length counts Unicode code points (Python
 * `len`), not UTF-16 units.
 *
 * Parity boundary (ADR-001): JS has one number type — a JSON `1.0` parses
 * to `1` and serializes as `"1"`, where Python emits `"1.0"`.
 */
import { AsyncLocalStorage } from "node:async_hooks";
import { compareCodePoints, pyFloat } from "./py.js";

export type Tokenizer = (text: string) => number;
export type Serializer = (value: unknown) => string;

export function codePointLength(text: string): number {
  let n = 0;
  for (const _ of text) n++;
  return n;
}

function heuristic(text: string): number {
  if (!text) return 0;
  return Math.max(1, Math.ceil(codePointLength(text) / 4));
}

/** Python-compatible canonical JSON (sort_keys, ensure_ascii=False, default=str). */
export function pythonJson(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) return value > 0 ? "Infinity" : value < 0 ? "-Infinity" : "NaN";
    // Python repr thresholds for floats ("1e-05", not JS's "0.00001");
    // integer-valued numbers collapse to int form (ADR-001 boundary 4a).
    return Number.isInteger(value) ? String(value) : pyFloat(value);
  }
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) {
    return "[" + value.map(pythonJson).join(", ") + "]";
  }
  if (value instanceof Set) {
    // Python json.dumps falls to default=str for sets; not JSON data — best effort.
    return JSON.stringify(String(value));
  }
  if (typeof value === "object") {
    // Python sorted() compares by code point; JS default string order is
    // UTF-16, which sorts astral-plane keys differently.
    const entries = Object.entries(value as Record<string, unknown>)
      .sort(([a], [b]) => compareCodePoints(a, b));
    return "{" + entries.map(([k, v]) => JSON.stringify(k) + ": " + pythonJson(v)).join(", ") + "}";
  }
  return JSON.stringify(String(value)); // default=str
}

let processTokenizer: Tokenizer | null = null;
let processSerializer: Serializer | null = null;

interface Scope { tokenizer?: Tokenizer; serializer?: Serializer }
const scopeStore = new AsyncLocalStorage<Scope>();

/** Process-wide default tokenizer; null resets to the built-in estimator. */
export function setTokenizer(tokenizer: Tokenizer | null): void {
  processTokenizer = tokenizer;
}

/** Process-wide default serializer; null resets to canonical JSON. */
export function setSerializer(serializer: Serializer | null): void {
  processSerializer = serializer;
}

/**
 * Context-local token accounting — safe for concurrent tenants. The JS
 * analog of Python's `tokens.scoped()`: overrides apply within `fn` and
 * every async continuation it spawns; other tasks are untouched.
 */
export function scoped<T>(overrides: Scope, fn: () => T): T {
  // merge with any enclosing scope so nested overrides compose, like
  // Python's two independent ContextVars; an explicit `undefined` means
  // "not provided", never "clobber the outer override"
  const merged: Scope = { ...scopeStore.getStore() };
  if (overrides.tokenizer !== undefined) merged.tokenizer = overrides.tokenizer;
  if (overrides.serializer !== undefined) merged.serializer = overrides.serializer;
  return scopeStore.run(merged, fn);
}

function activeTokenizer(): Tokenizer {
  return scopeStore.getStore()?.tokenizer ?? processTokenizer ?? heuristic;
}

function activeSerializer(): Serializer {
  return scopeStore.getStore()?.serializer ?? processSerializer ?? pythonJson;
}

/** True when the tokenizer in effect is the built-in estimator (enables
 * the closed-form truncate fast path, which relies on ceil(len/4)). */
export function isHeuristic(tokenizer?: Tokenizer | null): boolean {
  return (tokenizer ?? activeTokenizer()) === heuristic;
}

/** Render a value exactly the way token counting will price it. */
export function serialize(value: unknown): string {
  if (typeof value === "string") return value;
  return activeSerializer()(value);
}

/** Token cost of any serializable value (strings bypass the serializer). */
export function count(value: unknown, tokenizer?: Tokenizer | null): number {
  const tok = tokenizer ?? activeTokenizer();
  return tok(typeof value === "string" ? value : serialize(value));
}

File: csharp/Contexel/Tokens.cs
using System.Globalization;
using System.Text;

namespace Contexel;

/// <summary>
/// Token measurement — parity port of contexel's Python tokens module.
/// A record's cost is the token count of its serialized text; the
/// canonical serialization replicates Python's
/// json.dumps(value, ensure_ascii=False, default=str, sort_keys=True),
/// including the ", " / ": " separators, so all languages price identical
/// records identically. Lengths count Unicode code points (Python len).
/// Unlike the TS port, C# preserves int/float: a JSON 1.0 serializes as
/// "1.0" exactly like Python (ADR-002).
/// </summary>
public static class Tokens
{
    public sealed record Scope(Func<string, int>? Tokenizer = null,
                               Func<object?, string>? Serializer = null);

    private static readonly AsyncLocal<Scope?> ScopeLocal = new();
    private static Func<string, int>? _processTokenizer;
    private static Func<object?, string>? _processSerializer;

    public static int Heuristic(string text) =>
        text.Length == 0 ? 0 : Math.Max(1, (PyCompat.CodePointLength(text) + 3) / 4);

    /// <summary>Python-compatible canonical JSON (sort_keys,
    /// ensure_ascii=False, default=str). Hand-rolled: System.Text.Json
    /// escapes far more aggressively than json.dumps.</summary>
    public static string PythonJson(object? value)
    {
        var sb = new StringBuilder();
        WriteJson(sb, value);
        return sb.ToString();
    }

    private static void WriteJson(StringBuilder sb, object? value)
    {
        switch (value)
        {
            case null: sb.Append("null"); break;
            case bool b: sb.Append(b ? "true" : "false"); break;
            case long l: sb.Append(l.ToString(CultureInfo.InvariantCulture)); break;
            case int i: sb.Append(i.ToString(CultureInfo.InvariantCulture)); break;
            case double d:
                if (double.IsNaN(d)) sb.Append("NaN");
                else if (double.IsPositiveInfinity(d)) sb.Append("Infinity");
                else if (double.IsNegativeInfinity(d)) sb.Append("-Infinity");
                else sb.Append(PyCompat.PyFloat(d)); // "1.0", "1e-05" like Python
                break;
            case string s: WriteJsonString(sb, s); break;
            // generic type patterns are invariant (List<Rec> is not
            // List<object?>), so containers match structurally
            case System.Collections.IDictionary dict:
                sb.Append('{');
                bool first = true;
                foreach (var key in dict.Keys.Cast<object>().Select(k => (string)k)
                             .OrderBy(k => k, CodePointComparer.Instance))
                {
                    if (!first) sb.Append(", ");
                    first = false;
                    WriteJsonString(sb, key);
                    sb.Append(": ");
                    WriteJson(sb, dict[key]);
                }
                sb.Append('}');
                break;
            case System.Collections.IEnumerable seq:
                sb.Append('[');
                bool firstItem = true;
                foreach (var item in seq)
                {
                    if (!firstItem) sb.Append(", ");
                    firstItem = false;
                    WriteJson(sb, item);
                }
                sb.Append(']');
                break;
            default: // default=str fallback
                WriteJsonString(sb, Convert.ToString(value, CultureInfo.InvariantCulture) ?? "");
                break;
        }
    }

    private static void WriteJsonString(StringBuilder sb, string s)
    {
        sb.Append('"');
        foreach (char ch in s)
        {
            switch (ch)
            {
                case '"': sb.Append("\\\""); break;
                case '\\': sb.Append("\\\\"); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                case '\b': sb.Append("\\b"); break;
                case '\f': sb.Append("\\f"); break;
                default:
                    if (ch < 0x20)
                        sb.Append("\\u").Append(((int)ch).ToString("x4", CultureInfo.InvariantCulture));
                    else
                        sb.Append(ch); // ensure_ascii=False: non-ASCII stays raw
                    break;
            }
        }
        sb.Append('"');
    }

    private sealed class CodePointComparer : IComparer<string>
    {
        public static readonly CodePointComparer Instance = new();
        public int Compare(string? x, string? y) => PyCompat.CompareCodePoints(x!, y!);
    }

    /// <summary>Process-wide default tokenizer; null resets to the built-in estimator.</summary>
    public static void SetTokenizer(Func<string, int>? tokenizer) => _processTokenizer = tokenizer;

    /// <summary>Process-wide default serializer; null resets to canonical JSON.</summary>
    public static void SetSerializer(Func<object?, string>? serializer) => _processSerializer = serializer;

    /// <summary>Context-local overrides — safe for concurrent tenants
    /// (AsyncLocal, the .NET analog of contextvars). Nested scopes merge,
    /// so an inner serializer never clobbers an outer tokenizer.</summary>
    public static T Scoped<T>(Scope overrides, Func<T> fn)
    {
        var prev = ScopeLocal.Value;
        ScopeLocal.Value = Merge(prev, overrides);
        try { return fn(); }
        finally { ScopeLocal.Value = prev; }
    }

    public static async Task<T> ScopedAsync<T>(Scope overrides, Func<Task<T>> fn)
    {
        var prev = ScopeLocal.Value;
        ScopeLocal.Value = Merge(prev, overrides);
        try { return await fn().ConfigureAwait(false); }
        finally { ScopeLocal.Value = prev; }
    }

    private static Scope Merge(Scope? outer, Scope overrides) => new(
        overrides.Tokenizer ?? outer?.Tokenizer,
        overrides.Serializer ?? outer?.Serializer);

    private static Func<string, int> ActiveTokenizer() =>
        ScopeLocal.Value?.Tokenizer ?? _processTokenizer ?? Heuristic;

    private static Func<object?, string> ActiveSerializer() =>
        ScopeLocal.Value?.Serializer ?? _processSerializer ?? PythonJson;

    /// <summary>True when the tokenizer in effect is the built-in
    /// estimator (enables the closed-form truncate fast path).</summary>
    public static bool IsHeuristic(Func<string, int>? tokenizer = null) =>
        (tokenizer ?? ActiveTokenizer()) == (Func<string, int>)Heuristic;

    /// <summary>Render a value exactly the way token counting will price it.</summary>
    public static string Serialize(object? value) =>
        value is string s ? s : ActiveSerializer()(value);

    /// <summary>Token cost of any serializable value (strings bypass the serializer).</summary>
    public static int Count(object? value, Func<string, int>? tokenizer = null)
    {
        var tok = tokenizer ?? ActiveTokenizer();
        return tok(value is string s ? s : Serialize(value));
    }
}
